obspy/io/hypodd/pha.py

```diff
@@ -111,19 +128,128 @@ def _read_pha(filename, inventory=None, id_map=None, id_default='.{}..{}',
         substituted by station code and component.
     :param dict ph2comp: mapping of phases to components
         (default: {'P': 'Z', 'S': 'N'})
+    :param dict eventid_map: Desired mapping of hypodd event ids (dict values)
+        to event resource ids (dict keys).
+        The returned dictionary of the HYPODDPHA writing operation can be used.
+        By default, ids are not mapped.
     :param str encoding: encoding used (default: utf-8)
 
     :rtype: :class:`~obspy.core.event.Catalog`
     :return: An ObsPy Catalog object.
     """
+    if eventid_map is not None:
+        eventid_map = {v: k for k, v in eventid_map.items()}
     seed_map = _seed_id_map(inventory, id_map)
     with io.open(filename, 'r', encoding=encoding) as f:
         text = f.read()
-    events = [_block2event(block, seed_map, id_default, ph2comp)
+    events = [_block2event(block, seed_map, id_default, ph2comp, eventid_map)
               for block in text.split('#')[1:]]
     return Catalog(events)
 
 
+PHA1 = ('# {o.time.year} {o.time.month:>2} {o.time.day:>2} {o.time.hour:>2}'
+        ' {o.time.minute:>2} {o.time.second:>2}.{o.time.microsecond:06d}  '
+        '{o.latitude} {o.longitude} {depth}  {mag}  {he} {ve} {rms}'
+        '   {evid:>9}\n')
+PHA2 = '{p.waveform_id.station_code:6}  {relt:.4f}  {weight}  {p.phase_hint}\n'
+
+
+def _map_eventid(evid, eventid_map, used_ids, counter):
+    idpha = evid
+    if evid in eventid_map:
+        idpha = eventid_map[evid]
+        if not idpha.isdigit() or len(idpha) > 9:
+            msg = ('Invalid value in eventid_map, pha event id has to be '
+                   'digit with max 9 digits')
+            raise ValueError(msg)
+        return idpha
+    if not idpha.isdigit():
+        idpha = ''.join(char for char in idpha if char.isdigit())
+    if len(idpha) > 9:
+        idpha = idpha[:9]
+    while idpha == '' or idpha in used_ids:
+        idpha = str(counter[0])
+        counter[0] += 1
+    if idpha != evid:
+        eventid_map[evid] = idpha
```

Comment:
Could do with a final check that `len(idpha) <= 9` and an error being raised if it gets too big.

Reply:
For a catalog with more than a billion earthquakes? :smile: I can do that, but maybe a warnings is sufficient...

Reply:
Maybe just an assertion? That way it gets hit in the code coverage? It would be... unusual to have a catalog that large, but you could also inadvertently start `counter` close to the limit (which would also be... unusual).

Reply:
I think, I just leave it now as it is. I am pretty sure that the I9 limit will only be exceeded by the 10**10th event. It doesn't really matter as we will probably never see that large catalogs. (The warning I inserted is also pretty useless, but it's kind of correct.)